sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/DecorrelateInnerQuery.scala

```diff
@@ -0,0 +1,371 @@
+          // outer references and domain attributes when rewriting the domain joins.
+          // E.g. if the attribute a is changed to a1, the join condition a' <=> outer(a)
+          // will become a' <=> a1, and we can construct the aliases based on the condition:
+          // DomainJoin [a']        Join Inner
+          // +- InnerQuery     =>   :- InnerQuery
+          //                        +- Aggregate [a1] [a1 AS a']
+          //                           +- OuterQuery
+          val conditions = outerReferenceMap.map {
+            case (o, a) => EqualNullSafe(a, OuterReference(o))
+          }
+          (domainJoin, conditions.toSeq, outerReferenceMap)
+        }
+      } else {
+        plan match {
+          case Filter(condition, child) =>
+            val conditions = splitConjunctivePredicates(condition)
+            val (correlated, uncorrelated) = conditions.partition(containsOuter)
+            // Split the correlated predicates
+            val (equality, nonEquality) = correlated.partition(isEquality)
+            // Find outer references that can be substituted by attributes from the inner
+            // query using the equality predicates.
+            val equivalences = collectEquivalentOuterReferences(equality)
+            // Correlated predicates can be removed from the Filter's condition and used as
+            // join conditions with the outer query. However, if the results of the sub-tree
+            // is aggregated, only the correlated equality predicates can be used, because
+            // the inner query attributes from a non-equality predicate need to be preserved
+            // in both grouping and aggregate expressions, which can change the semantics
+            // of the plan and lead to incorrect results. Here is an example:
+            // Relations:
+            //   t1(c1, c2): [(1, 1)]
+            //   t2(c1, c2): [(1, 1), (2, 0)]
+            //
+            // Query:
+            //   SELECT * FROM t1 WHERE c1 = (SELECT MAX(c1) FROM t2 WHERE t1.c2 >= c2)
+            //
+            // Subquery plan transformation if non-equality predicates are used as join conditions:
+            //   Aggregate [max(c1)]                Aggregate [c2] [max(c1), c2]
+            //   +- Filter [outer(c2) >= c2]   =>   +- Relation [c1, c2]
+            //      +- Relation [c1, c2]
+            //
+            // Which will be rewritten to this query:
+            //   SELECT c1, c2 FROM t1 LEFT OUTER JOIN
+            //   (SELECT MAX(c1) m, c2 FROM t2 GROUP BY c2) s ON t1.c2 >= s.c2 WHERE c1 = m
+            //
+            // The result of the original query should be an empty set but the transformed
+            // query will output an incorrect result of (1, 1). The correct transformation
+            // is illustrated below:
+            //   Aggregate [max(c1)]                Aggregate [c2'] [max(c1), c2']
+            //   +- Filter [outer(c2) >= c2]   =>   +- Filter [c2' >= c2]
```

Comment:
Does it only apply to non-equality predicates? how about non-trivial equality predicates like `outer(c2) * 2 = c2`?

Reply:
I think the key is if the predicate can match more than one rows, and equality predicates usually only match one or zero rows.

However, it's not always the case, for example, `outer(c2) = substring(c2)` may match more than one rows.

Reply:
Since it's an optimization, shall we be conservative and do not pull out any predicate if the filter is under an aggregate?

Reply:
Brilliant finding! I've created a PR to disallow unsupported cases: https://github.com/apache/spark/pull/32179, and will update the code here.

Reply:
I checked the results for unsupported equality conditions with this framework and the results are correct.